### DISCUSSION TITLE
Is it possible to dump html into an element ie, ui.left_drawer()

### DISCUSSION THREAD

**Discussion by v479038280** [zauberzeug/nicegui#1955] [Q&A] [answered]
If is it possible to post html into an nicegui element. It may seems like this 

```py
with ui.left_drawer():
    ui.add_html_into_this_element('''
        <nav class="w3-sidebar w3-bar-block w3-small w3-hide-small w3-center">
          <!-- Avatar image in top left corner -->
          <img src="/w3images/avatar_smoke.jpg" style="width:100%">
          <a href="#" class="w3-bar-item w3-button w3-padding-large w3-black">
            <i class="fa fa-home w3-xxlarge"></i>
            <p>HOME</p>
          </a>
          <a href="#about" class="w3-bar-item w3-button w3-padding-large w3-hover-black">
            <i class="fa fa-user w3-xxlarge"></i>
            <p>ABOUT</p>
          </a>
          <a href="#photos" class="w3-bar-item w3-button w3-padding-large w3-hover-black">
            <i class="fa fa-eye w3-xxlarge"></i>
            <p>PHOTOS</p>
          </a>
          <a href="#contact" class="w3-bar-item w3-button w3-padding-large w3-hover-black">
            <i class="fa fa-envelope w3-xxlarge"></i>
            <p>CONTACT</p>
          </a>
        </nav>
    ''')
```
and here is the imported html look like
![微信截图_20231104141621](https://github.com/zauberzeug/nicegui/assets/25096045/84966e83-d376-407c-88d6-5a406924ac35)

In these html above ,there are  many CSS classes from third party, some may conflict with the nicegui CSS, could we choose which CSS library to use in the html head?
![微信截图_20231104142125](https://github.com/zauberzeug/nicegui/assets/25096045/e5993165-ab1b-4da1-96fc-938fb3ff9473)

**Comment by falkoschindler** [comment]
1. To add HTML to a page, you can use `ui.html`.
2. No, you can't simply choose a CSS library. NiceGUI needs Quasar to function properly. You can use `ui.run(tailwind=False)` to disable Tailwind though. But if you add W3.CSS, it probably conflicts with Quasar.

In summary: It is highly non-trivial to copy-and-paste some random HTML code into another page and to expect it to look nice, since it depends on so many global settings.

**Reply by v479038280** [reply]
It seems that the html we added are always at the beging of body ,if we could choose to put it to the end to body,some times the script added need to run after all the elements are created.

for example .I add the code below to the body,but the element  id #c31  cannot be found ,because it's excuted before the element was created  ,
and caught this error

> echarts.min.js:45  Uncaught TypeError: Cannot read properties of null (reading 'getAttribute')


```
<script type="text/javascript">
      var myChart = echarts.init(document.getElementById('c31'));

      var option = {
        title: {
          text: 'ECharts 入门示例'
        },
        tooltip: {},
        legend: {
          data: ['销量']
        },
        xAxis: {
          data: ['衬衫', '羊毛衫', '雪纺衫', '裤子', '高跟鞋', '袜子']
        },
        yAxis: {},
        series: [
          {
            name: '销量',
            type: 'bar',
            data: [5, 20, 36, 10, 10, 20]
          }
        ]
      };

      myChart.setOption(option);
    </script> 
```

**Reply by v479038280** [reply]
`console.log(document.getElementById('c31'))  
`
this print null in the console

**Reply by falkoschindler** [reply]
Crosspost: https://github.com/zauberzeug/nicegui/discussions/1960

**Comment by ed2050** [comment, marked as answer]
Yes it's difficult.  There are ways to make it a bit simpler though.

I did the same thing: took an existing html page and tried to recreate it in nicegui.  #1778 documents what I did.

External css was hard because of conflicts.  The page wouldn't even render with the external stylesheets applied.

Two things that helped:
1. Only copy the necessary parts of the external stylesheet.  90% of it related to layout, spacing, widths is unnecessary.  You can apply tailwind classes in the html block for most of that.  Parts to keep include colors and font stylings, those usually don't conflict with tailwind or quasar.
2. It's much easier to recreate layout structures in nicegui than to get the original html layout working.  With columns, rows, drawers, headers, tabs, etc, nicegui makes it quite easy to arrange content.  Easier than getting a tree of deeply nested divs from the original html to display properly.

I used these approaches in the example page, #1778.  The large chunks of html I kept were mostly paragraphs with embedded links, which translate well.

HTH,
